### PR TITLE
ref(cardinality): Recover buckets on cardinality limiter failure

### PR DESCRIPTION
**Pull request by Dav1dde**
Recover buckets on cardinality limiter error instead of discarding all

#skip-changelog

### CODE REVIEW

File: relay-cardinality/src/limiter.rs [hunk 1]
```diff
@@ -109,17 +109,24 @@ impl<T: Limiter> CardinalityLimiter<T> {
...
             let entries = items.iter().enumerate().filter_map(|(id, item)| {
                 Some(Entry::new(EntryId(id), item.to_scope()?, item.to_hash()))
             });
 
-            let rejections = self
-                .limiter
-                .check_cardinality_limits(organization, entries, self.config.cardinality_limit)?
-                .map(|rejection| rejection.id.0)
-                .collect::<BTreeSet<usize>>();
+            let rejections = self.limiter.check_cardinality_limits(
```
Comment (jjbayer) [inline relay-cardinality/src/limiter.rs]: nit

```suggestion
            let result = self.limiter.check_cardinality_limits(
```

File: relay-server/src/actors/processor.rs [hunk 2]
```diff
@@ -1449,21 +1450,29 @@ impl EnvelopeProcessorService {
         enable_cardinality_limiter: bool,
         _organization_id: u64,
         buckets: Vec<Bucket>,
-    ) -> Result<Box<dyn Iterator<Item = Bucket>>, relay_cardinality::Error> {
+    ) -> impl Iterator<Item = Bucket> {
         if !enable_cardinality_limiter {
-            return Ok(Box::new(buckets.into_iter()));
+            return Either::Left(buckets.into_iter());
```
Comment (iker-barriocanal) [inline relay-server/src/actors/processor.rs:1456]: IIUC `Left` is for buckets that are accepted and `Right` for those that are rate-limited. Is this correct? Is it worth documenting it?

**Review by iker-barriocanal — APPROVED** [review]

File: relay-server/src/actors/processor.rs [hunk 3]
```diff
@@ -1449,21 +1450,29 @@ impl EnvelopeProcessorService {
         enable_cardinality_limiter: bool,
         _organization_id: u64,
         buckets: Vec<Bucket>,
-    ) -> Result<Box<dyn Iterator<Item = Bucket>>, relay_cardinality::Error> {
+    ) -> impl Iterator<Item = Bucket> {
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs:1453]: Nice!

File: relay-server/src/actors/processor.rs [hunk 4]
```diff
@@ -1449,21 +1450,29 @@ impl EnvelopeProcessorService {
...
+                        "cardinality limiter failed"
+                    );
+
+                    Either::Left(buckets.into_iter())
+                }
+            };
         }
 
-        Ok(Box::new(buckets.into_iter()))
+        Either::<_, relay_cardinality::limiter::Accepted<_>>::Left(buckets.into_iter())
```
Comment (jjbayer) [inline relay-server/src/actors/processor.rs:1476]: Why does this need type annotations?

**Review by jjbayer — APPROVED** [review]

File: relay-server/src/actors/processor.rs (hunk 4 above)
Comment (Dav1dde) [reply relay-server/src/actors/processor.rs:1476]: For when we compile without `feature = processing`, then the `Right` side is never constructed -> error without annotations

File: relay-server/src/actors/processor.rs (hunk 2 above)
Comment (Dav1dde) [reply relay-server/src/actors/processor.rs:1456]: Exactly, but it doesn't have to be like that, `Left` and `Right` can contain [any iterator](https://docs.rs/itertools/0.12.0/itertools/enum.Either.html), it's just a helper to be able to return 2 distinct iterator types as one concrete type

**Pull request merged**